### DISCUSSION TITLE
Astro dev vs build

### DISCUSSION THREAD

**Discussion by suhaylmv** [swup/swup#680] [Help] [answered]
As I said, all the transitions work perfectly fine in local development server (npm run dev), although the navbar which was supposed to be fixed, doesn't follow the screen and remains at the top of the page for some reason. After I run "npm run build" and "npm run preview", there are no transitions and everything (including the navbar) works as before

Swup.js code:
```javascript 
import Swup from "swup";
import SwupSlideTheme from "@swup/slide-theme";
import SwupFadeTheme from '@swup/fade-theme';

const swup = new Swup({
  plugins: [new SwupSlideTheme()],
  options: {
    animationSelector: '[class*="transition-"]',
  },
theme: [new SwupFadeTheme()],
});
```

package.json code:
```json
{
  "name": "hidden",
  "type": "module",
  "version": "0.0.1",
  "scripts": {
    "dev": "astro dev",
    "start": "astro dev",
    "build": "astro build",
    "preview": "astro preview",
    "astro": "astro"
  },
  "dependencies": {
    "@astrojs/mdx": "^0.19.7",
    "@astrojs/solid-js": "^2.2.0",
    "@astrojs/tailwind": "^3.1.3",
    "@solid-primitives/intersection-observer": "^2.1.2",
    "@swup/fade-theme": "^1.0.5",
    "@swup/head-plugin": "^1.3.0",
    "@swup/slide-theme": "^1.0.5",
    "astro": "^2.6.3",
    "solid-icons": "^1.0.8",
    "solid-js": "^1.7.6",
    "swup": "^3.1.0",
    "tailwindcss": "^3.3.2"
  },
  "description": "``` npm create astro@latest -- --template minimal ```",
  "main": "index.js",
  "author": "",
  "license": "ISC"
}
```

**Comment by hirasso** [comment]
It's impossible to tell what is breaking your install during `astro build` without more information. Could you provide a link to a built version of your site online somewhere, so that we can have a look? 

The above swup initialization code is not fully correct, but that should not impact the outcome. Here is all you need to use the Fade Theme with swup (as per [the documentation](https://swup.js.org/themes/fade-theme/#usage)):

```js
import Swup from "swup";
import SwupFadeTheme from '@swup/fade-theme';

const swup = new Swup({
  theme: [new SwupFadeTheme()]
});
```

**Reply by suhaylmv** [reply]
Here's the repo of the site I was making: https://github.com/suhaylmv/amigos-del-retiro. You can clone it and run "npm run preview"

**Reply by hirasso** [reply]
Thanks for setting up the repo. It's missing some pages, only the main index.astro exists (also doesn't compile because the `events` collection is missing). For us to be able to really debug this, we will need more content.

**Reply by suhaylmv** [reply]
Can you explain more in detail what's wrong? I cloned my repo and I think there aren't any pages or collections missing, except for the about page (called "quienes somos") which I haven't made yet.

**Reply by hirasso** [reply]
When I run `npm run build`, I get the following error:

>  **error**   The collection events does not exist. Ensure a collection directory with this name exists.
  Hint:
    See https://docs.astro.build/en/guides/content-collections/ for more on creating collections.
  File:
    /{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/render/route-cache.js:21:27
  Code:
    20 |   let staticPaths = [];
    > 21 |   staticPaths = await mod.getStaticPaths({
         |                           ^
      22 |     paginate: generatePaginateFunction(route),
      23 |     rss() {
      24 |       throw new AstroError(AstroErrorData.GetStaticPathsRemovedRSSHelper);
  Stacktrace:
CollectionDoesNotExistError: The collection **events** does not exist. Ensure a collection directory with this name exists.
    at getCollection (file:///{my-test-folder}/amigos-del-retiro/dist/chunks/astro.ce2f02f9.mjs:3256:13)
    at Module.getStaticPaths$1 [as getStaticPaths] (file:///{my-test-folder}/amigos-del-retiro/dist/chunks/pages/_...page_.astro.8844dc20.mjs:143:26)
    at callGetStaticPaths (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/render/route-cache.js:21:27)
    at getPathsForRoute (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/build/generate.js:204:26)
    at generatePage (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/build/generate.js:185:23)
    at async generatePages (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/build/generate.js:117:7)
    at async staticBuild (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/build/static-build.js:85:7)
    at async AstroBuilder.build (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/build/index.js:124:5)
    at async AstroBuilder.run (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/build/index.js:156:7)
    at async build (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/core/build/index.js:37:3)
    at async runCommand (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/cli/index.js:171:14)
    at async cli (file:///{my-test-folder}/amigos-del-retiro/node_modules/astro/dist/cli/index.js:213:5)

**Reply by suhaylmv** [reply]
Seems very strange to me. You can see in the repo that the collection "events" exists:
![image](https://github.com/swup/swup/assets/64747225/23243011-f1b1-4a71-a9d4-a8b140408e5d)
Check if you have the same files as in the screenshot above

**Reply by hirasso** [reply]
Right! I have the `events` folder, as well. I don't know anything about Astro... does `npm run build` work on your end?

**Reply by suhaylmv** [reply]
Yeah, I cloned the repo and runned "npm run build", which worked. I'm using Arch Linux and I assume you're using Windows, but I don't think it should work any different. Maybe you could try cloning the repo in another directory, but honestly idk what could be causing your issue

**Reply by suhaylmv** [reply]
I've also shared the dist/ directory, so you should be able to just run "npm run preview", have you tried it? Try also running "npm run dev"

**Reply by hirasso** [reply]
I'm on macOS. Have you tried to pull in the repo you've published and install that? There is also a proper `.gitignore` missing. That should contain at least the following:

```gitignore
node_modules
.DS_Store
```

`npm run preview` also gives me a missing collection error in the browser, when I visit `/eventos/`:

![image](https://github.com/swup/swup/assets/869813/74b68410-45f6-4879-9310-3bb00b7e1fea)

**Reply by suhaylmv** [reply]
What do you mean by installing? I forgot to create .gitignore, maybe I'll add it later. I suggest deleting the events page, it's not relevant to the swup problem anyways

**Reply by hirasso** [reply]
> What do you mean by installing?

I want you to test if this works on your end:

```shell
git clone git@github.com:suhaylmv/amigos-del-retiro.git
cd amigos-del-retiro
npm run build
```
>  I suggest deleting the events page, it's not relevant to the swup problem anyways

I don't know anything about Astro or it's setup. Basically what I would need from you is a very simple two-page astro app, so that we can debug swup and don't have to worry about Astro.

> maybe I'll add it later. 

Thanks! We'll wait until the repo is fully functional.

**Reply by suhaylmv** [reply]
Got this: 
```bash
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.
```
Idk why, I have very little experience with git tbh. Creating a simpler astro project seems a better idea, I'll send you a link when I finish

**Reply by suhaylmv** [reply]
https://github.com/suhaylmv/test
Install dependencies, run "npm run build" and "npm run preview"

**Reply by hirasso** [reply]
Thanks for the reduced test site! I was able to get it to work. Two things to keep in mind:

**1.** You defined a custom transition class `.transition-fade`, with your own custom CSS for that class: 

```css
/* global.css */
.transition-fade {
  transition: 0.4s;
  opacity: 1;
}
html.is-animating .transition-fade {
  opacity: 0;
}

```

If you do this, you don't have to use a theme. It might even conflict with your custom transition. So your Swup initialization code should simply look like this:

```js
import Swup from "swup";
const swup = new Swup();
```

**2.** The `#swup` element needs to be present on _every on of your pages_. On `BlogPost.astro`, for example, it was missing. Also, you should have the `class="transition-fade"` on every page where you want to have a transition. Your `<main>` element on `BlogPost.astro` should look like this: `<main id="swup" class="transition-fade">`.

We recommend you invest some time reading about the basics of swup on the [official docs](https://swup.js.org/). That will save you a lot of trouble down the road.

**Reply by hirasso** [reply]
Weird thing is, not even with the starter kit the collections are working for me... But that's an astro problem, not a swup problem 😄

**Reply by suhaylmv** [reply]
Wow, didn't know about that. The thing is I couldn't understand the docs at all and I started searching for random tutorials, mixed up some concepts and ended up writing something that doesn't make any sense. 
I've read the docs like 10 times and I still can't get the slide transition working, either the docs are a bit poorly written or I'm being stupid. Could you please check out again the repo and tell me what's wrong? https://github.com/suhaylmv/test

**Reply by hirasso** [reply]
I have solved my [issue with astro](https://github.com/withastro/astro/issues/7406#issuecomment-1596226682) and was finally able to properly test your site! 

Turns out that there is indeed an error in the documentation about using themes!! The docs wrongly say that the option for using a theme should be `theme`, while instead the theme should simply be registered like any other swup plugin under the `plugins` property 🤦‍♀️ 

So this is what you should do to make this work:

```js
const swup = new Swup({
  plugins: [new SwupSlideTheme()] // instead of theme: ...
});
```

You should then also remove all your custom swup-targeted css in `MainLayout.astro`, the theme will take care of that for you:

```diff
<style>
-  .transition-fade {
-    transition: 0.4s;
-    opacity: 1;
-  }

-  html.is-animating .transition-fade {
-    opacity: 0;
-  }
</style>
```
Hope this helps. I will create a PR for fixing the false information on the docs right away.

**Reply by suhaylmv** [reply]
Thanks. I'm glad you fixed your issue with Astro and noticed an error in documentation. I did exactly what you said, and the same happens as before, transitions work in "npm run dev", but not after building the site. You can try it yourself: https://github.com/suhaylmv/test

**Comment by daun** [comment]
Could it be related to how you're including the JS? It's included as `<script type=module>` and importing `swup` directly — not sure if Astro creates an import map for you or if it just bundles everything.

**Comment by hirasso** [comment, marked as answer]
I gave it a try. Turns out astro behaves differently during `dev` vs `build`. The top-level file `swup.js` was not being copied to the `dist` folder during `npm run build`. I was able to get it to work in both environments by:

1. Removing the top-level file `swup.js`
2. Removing the `<script type="module" src="/swup.js"></script>` on every page layout
3. Adding the script as an astro script inside `Footer.astro` (so it's being loaded on every page):

```html
---
const today = new Date();
---

<footer>
	&copy; {today.getFullYear()} YOUR NAME HERE. All rights reserved.
</footer>
<style>
	footer {
		padding: 25px;
		text-align: center;
	}
</style>
<script>
	import Swup from "swup";
	// @ts-ignore
	import SwupSlideTheme from "@swup/slide-theme";

	const swup = new Swup({
		plugins: [new SwupSlideTheme()],
	});
</script>

```
It's an astro thing. More information: https://docs.astro.build/en/guides/client-side-scripts/

**Reply by suhaylmv** [reply]
Great solution! Thank you so much for your time and effort. Works perfectly during build. Now there's only one problem/side effect remaining, which I had since the beginning, the navbar is not fixed to the top. I think it could be related to this https://swup.js.org/other/common-issues/#scroll-control-offset, but that piece of information lacks context and I couldn't understand it. Once I have more experience with this great library I might consider contributing by improving the docs, or translating them to Spanish or Russian )
